src/main/java/org/broadinstitute/hellbender/tools/exome/cnlohcaller/CNLOHCaller.java

```diff
@@ -0,0 +1,752 @@
+            for (int mi = 0; mi < mVals.length; mi++) {
+                for (int ni = 0; ni < nVals.length; ni++) {
+                    result[rhoi][mi][ni] /= total;
+                }
+            }
+        }
+
+        return result;
+    }
+
+    protected double[] calc_E_ln_phi_k(final double E_alpha, final double[] eZskSummedOverSegs) {
+
+        final double eZskSummedOverAll = MathUtils.sum(eZskSummedOverSegs);
+
+        final double[] result = new double[eZskSummedOverSegs.length];
+        final int k = eZskSummedOverSegs.length;
+
+        // Initialize all pseudocounts to 1, except for index 0, which is 20;
+        //  This artificially increases the odds for a rho = 0.
+        final RealVector pseudocounts = new ArrayRealVector(eZskSummedOverSegs.length);
+        pseudocounts.set(1.0);
+        pseudocounts.setEntry(0, 20.0);
+        final double sumPseudocounts = MathUtils.sum(pseudocounts.toArray());
+
+        for (int i=0; i < result.length; i++) {
+            final double term1 = Gamma.digamma(E_alpha/k + pseudocounts.getEntry(i) + eZskSummedOverSegs[i]);
+            final double term2 = Gamma.digamma(E_alpha + sumPseudocounts + eZskSummedOverAll);
+            result[i] = term1 - term2;
+        }
+
+        return result;
+    }
+
+    protected double[] calcPis(final double[] eVsm, final double[] eWsn) {
+
+        // Sum (and normalize) the responsibilities for each possible value of the absolute copy numbers.
+
+        double total = 0;
+        for (int i = 0; i < eVsm.length; i++) {
+            total += eVsm[i] + eWsn[i];
+        }
+
+        final double[] result = new double[eVsm.length];
+        for (int i = 0; i < result.length; i++) {
+            result[i] = (eVsm[i] + eWsn[i]) / total;
+        }
+        return result;
+    }
+
+    protected double calcEAlpha(final double[] E_ln_phi_k, final int minCN, final int maxCN) {
+        final QAlphaUnivariateFunction qAlpha = new QAlphaUnivariateFunction(E_ln_phi_k);
+        final AQAlphaUnivariateFunction aQAlpha = new AQAlphaUnivariateFunction(E_ln_phi_k);
+        final BaseAbstractUnivariateIntegrator integrator = new SimpsonIntegrator();
+
+        final double numerator = integrator.integrate(10, aQAlpha, minCN, maxCN);
+        final double denominator = integrator.integrate(10, qAlpha, minCN, maxCN);
+
+        // If denominator and numerator are both zero, we want to return zero.  But that simplifies to the
+        //  numerator being zero.
+        if (numerator == 0) {
+            return 0;
+        } else {
+            return numerator/denominator;
+        }
+    }
+
+    private double[] calcNewRhos(final List<ACNVModeledSegment> segments,
+                               final List<double[][][]> eZskVsmWsnBySeg, final SortedMap<Integer, Double> pis,
+                               final double lambda, final double[] rhos, final JavaSparkContext ctx) {
+
+        final int[] mVals = pis.keySet().stream().mapToInt(i->i).toArray();
```

Comment:
Treating this as a `Map` is making the code more cumbersome than it needs to be. . .